boot.kernel.features: add nixos option for setting kernel features

bendlas opened the pull request:
###### Motivation for this change

The motivation is submitted in the form of additional commits on this PR:
- provide options for installing `systemtap` and `criu`; both profit from / require custom kernel config
- optionally enabling `CONFIG_DEBUG_INFO`, for use with systemtap
- optionally enabling `CONFIG_CHECKPOINT_RESTORE`, which is required by `criu`

###### Review requested

- Is it done the right way?
- What about the section re-enabling flags, that, for some reason, got disabled by `CONFIG_EXPERT` (`EXPERT` is needed for `CHECKPOINT_RESTORE`) https://github.com/NixOS/nixpkgs/pull/34791/commits/d86f2529529f27f19d72e85830748c4a62e9b937#diff-ad3f665c58e343da48fc9981dde5afaeR690

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

bendlas committed:
boot.kernel.features: add nixos option for setting kernel features

this allows setting the features attribute for
pkgs/os-specific/linux/kernel/common-config.nix

bendlas committed:
programs.criu: add nixos option for installing criu + kernel flags

bendlas committed:
linuxPackages.kernel: add feature.debug flag for enabling DEBUG_INFO

bendlas committed:
programs.systemtap: add nixos option for installing systemtap

also enables debug feature on kernel

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=433f9932-5faf-44ae-86a4-37b1ab619e94)

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/lxvqgcmllzd2if24js1nyd43r3msczlr-linux-4.9.80
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=4ff65878-d5c9-4acd-b60a-43a1ae026407)

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/pci/hotplug/shpchp_hpc.o
  LD [M]  drivers/pci/hotplug/shpchp.o
  CC [M]  drivers/pci/hotplug/acpiphp_ibm.o
  CC      drivers/pci/pcie/aspm.o
  CC      drivers/pci/pcie/portdrv_core.o
  CC      drivers/pci/pcie/portdrv_pci.o
  CC      drivers/pci/pcie/portdrv_bus.o
  CC      drivers/pci/pcie/portdrv_acpi.o
building of ‘/nix/store/70w8ifg3wvndk7c47c4gnrkb6vghj8hf-linux-4.9.80.drv’ timed out after 3600 seconds
error: build of ‘/nix/store/70w8ifg3wvndk7c47c4gnrkb6vghj8hf-linux-4.9.80.drv’ failed
```
</p></details>

bendlas commented on nixos/modules/system/boot/kernel.nix:
```diff
@@ -21,6 +23,15 @@ in
+      default = {};
+      example = literalExample "{ debug = true; }";
+      description = ''
+        This option allows to enable or certain kernel features.
```
Thanks!

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=7daeb1d2-14fb-4be0-bb35-a4e403325595)

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/lxvqgcmllzd2if24js1nyd43r3msczlr-linux-4.9.80
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=fe06bda6-1994-462b-9bee-19521444ca53)

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
/nix/store/d962jhp5pqaywvcw9bmb5hgfs5sb96r5-linux-4.9.80
```
</p></details>

NeQuissimus commented:
@GrahamcOfBorg test kernel-lts

NeQuissimus commented:
@GrahamcOfBorg test kernel-latest

NeQuissimus commented:
@GrahamcOfBorg test kernel-copperhead

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=0bfb73f7-0308-40ca-aad4-22ff23d6e52c)

<details><summary>Partial log (click to expand)</summary><p>

```
while evaluating the option `boot.kernelPackages':
while evaluating the attribute 'isDefined' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:344:5:
while evaluating 'filterOverrides' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:419:21, called from /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:328:18:
while evaluating 'concatMap' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/lists.nix:102:18, called from /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:425:8:
while evaluating 'concatMap' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/lists.nix:102:18, called from /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:323:17:
while evaluating anonymous function at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:323:28, called from undefined position:
while evaluating 'dischargeProperties' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:386:25, called from /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:324:62:
while evaluating the attribute 'value' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/lib/modules.nix:216:48:
while evaluating the attribute 'config.kernelPackages' at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/nixos/tests/kernel-copperhead.nix:9:7:
attribute 'linuxPackages_hardened_copperhead' missing, at /var/lib/gc-of-borg/nix-test-rs-1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-aarch64-community-1/nixos/tests/kernel-copperhead.nix:9:29
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=abaf0903-40b0-423d-a447-6cea67ea4b69)

<details><summary>Partial log (click to expand)</summary><p>

```
machine# [    7.467137] dhcpcd[763]: eth0: IAID 00:12:34:56
machine# [    7.467784] dhcpcd[763]: eth0: adding address fe80::5054:ff:fe12:3456
machine# [    7.681703] input: ImExPS/2 Generic Explorer Mouse as /devices/platform/i8042/serio1/input/input4
machine: exit status 0
test script finished in 8.57s
cleaning up
killing machine (pid 593)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/wc33p63rxdbgzffdbp0z4dps7wiy519l-vm-test-run-kernel-lts
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=758e29d6-78a1-4e51-984f-83c78bfae106)

<details><summary>Partial log (click to expand)</summary><p>

```
while evaluating the option `boot.kernelPackages':
while evaluating the attribute 'isDefined' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:344:5:
while evaluating 'filterOverrides' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:419:21, called from /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:328:18:
while evaluating 'concatMap' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/lists.nix:102:18, called from /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:425:8:
while evaluating 'concatMap' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/lists.nix:102:18, called from /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:323:17:
while evaluating anonymous function at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:323:28, called from undefined position:
while evaluating 'dischargeProperties' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:386:25, called from /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:324:62:
while evaluating the attribute 'value' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/lib/modules.nix:216:48:
while evaluating the attribute 'config.kernelPackages' at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/nixos/tests/kernel-copperhead.nix:9:7:
attribute 'linuxPackages_hardened_copperhead' missing, at /home/borg/borg-rs/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/7c6f434c-buildbox/nixos/tests/kernel-copperhead.nix:9:29
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=c041e3be-2de4-4eeb-91ca-13ed3483c6b1)

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation ‘/nix/store/yvyzrr7alliivyy1shvlb3a1p1l9h8y5-etc.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/lbggcwgm9wqyr9yzdkp4ylnsc8sm0h12-stage-1-init.sh.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/ihp2f95zwgshkz8f3jp32yqfljzgrbsk-initrd.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/sb2l8xsvfish2n7j8bhkw76l6gr0qbpi-nixos-system-machine-18.03.git.69f14c0.drv’: 5 dependencies couldn't be built
cannot build derivation ‘/nix/store/fd4k0rp27qchfn5v96xii159i6fgprsi-closure-info.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/shx2lrly2bjlyjpa8kr2hmqmlshbx4hw-run-nixos-vm.drv’: 2 dependencies couldn't be built
cannot build derivation ‘/nix/store/lcnrdiqlzy1gc17qq32d1d6qwd7syvy3-nixos-vm.drv’: 2 dependencies couldn't be built
cannot build derivation ‘/nix/store/6d7ci4kr2mvb70x89mc8q5wv5ys8adbl-nixos-test-driver-kernel-latest.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/c89yxzn26df9c7mp7jj8y5hqhbzgym9m-vm-test-run-kernel-latest.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/c89yxzn26df9c7mp7jj8y5hqhbzgym9m-vm-test-run-kernel-latest.drv’ failed
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=fa956df1-9010-46e5-9bd0-d8001bdc4e21)

<details><summary>Partial log (click to expand)</summary><p>

```
machine: exit status 1
syncing
machine: running command: sync
machine: exit status 0
test script finished in 14.27s
cleaning up
killing machine (pid 627)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/pknym1flwk8qbyrf2ry5r29v7fg9zvnx-vm-test-run-kernel-lts
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34791&attempt_id=d7701782-a73f-422d-952b-477564b1db0b)

<details><summary>Partial log (click to expand)</summary><p>

```
machine: exit status 1
syncing
machine: running command: sync
machine: exit status 0
test script finished in 14.61s
cleaning up
killing machine (pid 627)
vde_switch: EOF on stdin, cleaning up and exiting
vde_switch: Could not remove ctl dir '/build/vde1.ctl': Directory not empty
/nix/store/31wwxqiwbzq12jjq2dmp3ava1ak75k7v-vm-test-run-kernel-latest
```
</p></details>

bendlas committed:
nixos: make `boot.kernel.features` internal

This isn't useful as public API. It should be used by options to
activate kernel features for use with specific programs.

bendlas reviewed (commented):
features should be introduced to the kernel via the existing override in `kernelPackages`

bendlas commented:
I redid the override, hopefully fixing it.
I also set the option to internal, because it's not clear to me how to properly document it.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34791&attempt_id=b7b4d601-d087-494e-b8b8-d05813c3644c)

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/zhmb4jv153l3glh1dxb8vvnhr96g2ycz-linux-4.9.81
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34791&attempt_id=b51e62ec-5ca4-44ad-a1f5-3780e67f1284)

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/dm.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/hw.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/led.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/mac.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/phy.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/rf.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/sw.o
  CC [M]  drivers/net/wireless/realtek/rtlwifi/rtl8192cu/table.o
building of ‘/nix/store/ap7p0i32zmdirrxj68nwh3spbb31i9vk-linux-4.9.81.drv’ timed out after 3600 seconds
error: build of ‘/nix/store/ap7p0i32zmdirrxj68nwh3spbb31i9vk-linux-4.9.81.drv’ failed
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34791&attempt_id=e40ab88a-b293-4337-aa3e-052df7ce23e4)

<details><summary>Partial log (click to expand)</summary><p>

```
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
wrong ELF type
cannot find section .dynamic
/nix/store/zhmb4jv153l3glh1dxb8vvnhr96g2ycz-linux-4.9.81
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34791&attempt_id=36ca3c72-6f3e-47b2-b482-2eade8f21bb1)

<details><summary>Partial log (click to expand)</summary><p>

```
  CC [M]  drivers/net/wireless/marvell/mwifiex/11n.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/11ac.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/11n_aggr.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/11n_rxreorder.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/scan.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/join.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/sta_ioctl.o
  CC [M]  drivers/net/wireless/marvell/mwifiex/sta_cmd.o
building of ‘/nix/store/ap7p0i32zmdirrxj68nwh3spbb31i9vk-linux-4.9.81.drv’ timed out after 3600 seconds
error: build of ‘/nix/store/ap7p0i32zmdirrxj68nwh3spbb31i9vk-linux-4.9.81.drv’ failed
```
</p></details>
